src/gt4py/stencil_object.py

```diff
@@ -118,17 +128,14 @@ class StencilObject(abc.ABC):
     _gt_id_: str
     definition_func: Callable[..., Any]
 
+    _seen_domain_origin: Dict[int, Tuple[Tuple[int, ...], Dict[str, Tuple[int, ...]]]] = {}
+    """Stores domain/origin pairs that have been used by hash."""
+
     def __new__(cls, *args, **kwargs):
         if getattr(cls, "_instance", None) is None:
             cls._instance = object.__new__(cls)
         return cls._instance
 
-    def __setattr__(self, key, value) -> None:
-        raise AttributeError("Attempting a modification of an attribute in a frozen class")
-
-    def __delattr__(self, item) -> None:
-        raise AttributeError("Attempting a deletion of an attribute in a frozen class")
```

Comment:
This is implemented at the expense of freezing the StencilObject class. Does anyone see this as a problem? We could potentially do this in a wrapper class a la FrozenStencil.